exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,838 @@
+    /// }
+    /// ```
+    pub fn indices(&self) -> SparceListIndexKeys {
+        SparceListIndexKeys { base_iter: self.base.iter_from(&(), &0u64) }
+    }
+
+    /// Returns an iterator over the values of the 'SparceListIndex'. The iterator element type is
+    /// V.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.extend([1, 2, 3, 4, 5].iter().cloned());
+    ///
+    /// for val in index.values() {
+    ///     println!("{}", val);
+    /// }
+    /// ```
+    pub fn values(&self) -> SparceListIndexValues<V> {
+        SparceListIndexValues { base_iter: self.base.iter_from(&(), &0u64) }
+    }
+
+    /// Returns an iterator over the list starting from the specified position. The iterator
+    /// element type is (u64, V).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.extend([1, 2, 3, 4, 5].iter().cloned());
+    /// index.remove(3);
+    ///
+    /// for val in index.iter_from(3) {
+    ///     println!("{:?}", val);
+    /// }
+    /// ```
+    pub fn iter_from(&self, from: u64) -> SparseListIndexIter<V> {
+        SparseListIndexIter { base_iter: self.base.iter_from(&(), &from) }
+    }
+}
+
+
+impl<'a, V> SparseListIndex<&'a mut Fork, V>
+where
+    V: StorageValue,
+{
+    fn set_size(&mut self, size: SparseListSize) {
+        self.base.put(&(), size);
+        self.size.set(Some(size));
+    }
+
+    /// Appends an element to the back of the list.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.push(1);
+    /// assert!(!index.is_empty());
+    /// ```
+    pub fn push(&mut self, value: V) {
+        let mut size = self.size();
+        self.base.put(&size.capacity, value);
+        size.capacity += 1;
+        size.length += 1;
+        self.set_size(size);
+    }
+
+    /// Removes the last element from the list and returns it, or None if it is empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(None, index.pop());
+    ///
+    /// index.push(1);
+    /// assert_eq!(Some(1), index.pop());
+    /// ```
+    pub fn pop(&mut self) -> Option<V> {
```

Comment:
This method is not very obvious since it uses the number of non-empty items to index in a sparse list. In the first place, this means that it extracts the element from the middle of the list, and secondly, it can return `None` in a situation where the index is not empty yet.

I think that we should leave only 2 methods:

- `pop` which works exactly like `pop_front` (and of course its behavior should be specified in the docs)
- `push` which works like `push_back`